[Gowri Bhat] iP

gowribhat opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

gowribhat committed:
Level-1

gowribhat committed:
contains list elements

gowribhat committed:
task class added with marking feature

gowribhat committed:
subclasses for Task added

gowribhat committed:
changed format of files and impleted exceptions

gowribhat committed:
Tests added and more errors handled

gowribhat committed:
delete feature added

gowribhat committed:
enums added

gowribhat committed:
changes to duke formatting

gowribhat committed:
private helpers

gowribhat committed:
minor formatting

gowribhat committed:
code clean up - error handling and helpers

gowribhat committed:
updated test

gowribhat committed:
Store deadline and event dates as LocalDate in Task objects

gowribhat committed:
Update Duke.java

gowribhat committed:
Add save and retrieve data feature

gowribhat committed:
Merge branch 'master' of https://github.com/gowribhat/ip into branch-Level-8

gowribhat committed:
Add more classes and organize into packages

gowribhat committed:
Add data folder to .gitignore

gowribhat committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

gowribhat committed:
Add JUnit tests

gowribhat committed:
Reorganize code

gowribhat committed:
Update JUnit tests

gowribhat committed:
Add JavaDoc

gowribhat committed:
Update code to comply with a coding standard

gowribhat committed:
Add Find feature

gowribhat committed:
Merge branch 'branch-A-JavaDoc'

gowribhat committed:
Merge branch 'branch-A-CodingStandard' and resolve conflicts

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/task/Task.java
#	src/test/java/duke/task/DeadlineTest.java

gowribhat committed:
Merge branch 'branch-Level-9' and resolve conflicts

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/task/Task.java

gowribhat committed:
Add more JavaDoc and reorganize packages

gowribhat committed:
Update JavaDoc

gowribhat committed:
Add JavaDoc in Ui.java

gowribhat committed:
Add checkStyle to detect coding style violations

gowribhat committed:
Fix checkstyle

gowribhat committed:
Fix styling in Ui.java

gowribhat committed:
Merge branch 'branch-A-CheckStyle'

Eugenetanwl3881 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,140 @@
+     */
+    private void parseCreateTask(String[] splitInput, TaskList tasks) throws DukeException {
+        int len = splitInput.length;
+        if (len == 0) {
+            throw new DukeException("No input!");
+        } else if (splitInput[0].equals("todo")) {
+            tasks.editTaskList("todo", getTaskField(splitInput, 1, len), "");
+        } else if (splitInput[0].equals("deadline")) {
+            int index = getDateIndex(splitInput);
+            String taskDescription = getTaskField(splitInput, 1, index);
+            String date = getTaskField(splitInput, index + 1, len);
+            tasks.editTaskList("deadline", taskDescription, date);
+        } else if (splitInput[0].equals("event")) {
+            int index = getDateIndex(splitInput);
+            String taskDescription = getTaskField(splitInput, 1, index);
+            String date = getTaskField(splitInput, index + 1, len);
+            tasks.editTaskList("event", taskDescription, date);
+        } else {
+            throw new DukeException("Duke: OH NO!!! I'm sorry, but I don't know what that means :-(");
+        }
```
Do you think the code can be made more readable/obvious and clean by using enums and a switch statement?

Eugenetanwl3881 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,140 @@
+     * @throws DukeException if command does not follow the format
+     */
+    public void parse(String userInput, TaskList tasks) throws DukeException {
+        String[] splitInput = userInput.split(" ");
+        if (userInput.equals("bye")) {
+            isExit = true;
+        } else if (userInput.equals("list")) {
+            Ui.printTaskList(tasks);
+        } else if(splitInput[0].equals("find")) {
+            parseFind(splitInput, tasks);
+        } else if (splitInput[0].equals("mark")) {
+            parseEdit("mark", splitInput, tasks);
+        } else if (splitInput[0].equals("unmark")) {
+            parseEdit("unmark", splitInput, tasks);
+        } else if (splitInput[0].equals("delete")) {
+            parseEdit("delete", splitInput, tasks);
+        } else if (splitInput[0].equals("delete")) {
+            throw new DukeException("Duke: To delete tasks, indicate the index of the task using an integer!");
+        } else {
+            parseCreateTask(splitInput, tasks);
```
I like your method namings. It is in verb form and easy to understand what the parser is doing. Good job! 😄

Eugenetanwl3881 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,65 @@
+     * @param tasks list of stored tasks
+     */
+    public void saveData(TaskList tasks) {
+        String csv = tasks.toCsv();
```
Good job that you remembered to make _toCsv_ method camelCase. People tend to forget that and may write Csv in all caps👍

Eugenetanwl3881 commented on src/main/java/duke/Ui.java:
````````````````````````````````````````````````````````````````````diff
@@ -0,0 +1,105 @@
+    private static final String LINE = "```````````````````````````````````````````````````````````````````";
+    private static final String GREET_WELCOME = "Hello there! I am\n" + LOGO
+            + "\nyour personal task tracking assistant!\nWhat can I do for you today?\n";
+    private static final String GREET_EXIT = "Bye. Hope to see you again soon!\n";
````````````````````````````````````````````````````````````````````
Is it necessary to make this a constant? Or do you we can just have them in the methods since they are only used once? 🤔

nitant-p replied:
I think it makes sense to expect it to be used more than once in the future as the project develops so no harm in making it constant 🤔

Eugenetanwl3881 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+ * Represents a list of <code>Task</code> objects.
+ */
+public class TaskList {
+    private LinkedList<Task> tasks = new LinkedList<>();
```
Good naming that you wrote tasks in plural !

Eugenetanwl3881 commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,31 @@
+    private TaskList testList = new TaskList(testDataFile);
+
+    @Test
+    public void toCsv_taskList_taskListCsvReturned() throws IOException {
```
Clear test naming 👍

Eugenetanwl3881 reviewed (commented):
Code quality overall is good! Method names and variable names were named appropriately and cased properly. You may wish to make some parts of the code more "obvious" like parts wjhere switch statements can replace many lines of "if-else". Everything else LGTM. Good job 👍

nitant-p commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,50 @@
+     */
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
```
Nice, the boolean variable follows the naming convention 😄

nitant-p commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,65 @@
+ * Deals with loading <code>Tasks</code> from storage file and saving <code>Tasks</code> in the file.
+ */
+public class Storage {
+    private static final String FILENAME = "saved-tasks.txt";
```
Constant is correctly named and is at the top since it's a static variable 👍

nitant-p commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class Storage {
+    private static final String FILENAME = "saved-tasks.txt";
+    private final File dataFile;
```
Shouldn't this constant be in upper case? 🤔

nitant-p commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,105 @@
+        this.scanner.close();
+        System.out.println(GREET_EXIT + LINE);
+    }
+}
```
Methods are clearly names so it's easier to follow your program 😄

nitant-p commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+     * @param keyword word being searched for
+     * @return true if and only if keyword in a substring of the task description
+     */
+    public boolean contains(String keyword) {
```
Not sure about this but what about renaming the method name to doesContain or isContained/ Containing? 🤔 Sounds a bit odd though

nitant-p commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,31 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
I think the static import should be at the top of the list 🤔

nitant-p reviewed (commented):
Overall, I think your code follow the coding standards very closely and it has great readability! 😄  I just have a suggestions for the naming and ordering of some variables.

gowribhat committed:
Add basic GUI for Duke

gowribhat committed:
Fix minor GUI bugs and reorganize controllers into a package

gowribhat committed:
Clean up code

gowribhat committed:
Merge branch 'branch-Level-10'

gowribhat committed:
Add commands and enums

gowribhat committed:
Organize files into packages

gowribhat committed:
Add assertions

gowribhat committed:
Improve code quality

Making the code readable and simple is important. Better code quality
allows debugging to be smoother and provides clarity for others reading
the code.

gowribhat committed:
Merge pull request #1 from gowribhat/branch-A-Assertions

Add assertions

gowribhat committed:
Merge branch 'master' of https://github.com/gowribhat/ip into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Parser.java

gowribhat committed:
Merge pull request #2 from gowribhat/branch-A-CodeQuality

Improve code quality

gowribhat committed:
Add Notes feature

Users can add and delete notes which will be displayed separately
from other tasks.

Currently only Todos, Deadlines and Events can be added.

gowribhat committed:
Add javadoc and fix checkstyle errors

gowribhat committed:
Set up CI

gowribhat committed:
Update gradle.yml

gowribhat committed:
Suppress checkstyle

gowribhat committed:
Fix bug in build.gradle

gowribhat committed:
Add more JUnit Tests

gowribhat committed:
Correct code quality error

gowribhat committed:
Add basic User Guide

gowribhat committed:
Update README.md

Include hyperlinks for easier navigation through the user guide

gowribhat committed:
Merge branch 'branch-A-UserGuide'

gowribhat committed:
Update README.md

gowribhat committed:
Update exception messages

gowribhat committed:
Merge branch 'master' of https://github.com/gowribhat/ip

gowribhat committed:
Change exception message

gowribhat committed:
Update UiTest.java

gowribhat committed:
Add link to download Duke
